flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/MergingSharedSlotProfileRetrieverFactory.java

```diff
@@ -40,38 +39,37 @@
 
     private final Function<ExecutionVertexID, AllocationID> priorAllocationIdRetriever;
 
+    private final Supplier<Set<AllocationID>> reservedAllocationIdsRetriever;
+
     MergingSharedSlotProfileRetrieverFactory(
             SyncPreferredLocationsRetriever preferredLocationsRetriever,
-            Function<ExecutionVertexID, AllocationID> priorAllocationIdRetriever) {
+            Function<ExecutionVertexID, AllocationID> priorAllocationIdRetriever,
+            Supplier<Set<AllocationID>> reservedAllocationIdsRetriever) {
         this.preferredLocationsRetriever = Preconditions.checkNotNull(preferredLocationsRetriever);
         this.priorAllocationIdRetriever = Preconditions.checkNotNull(priorAllocationIdRetriever);
+        this.reservedAllocationIdsRetriever =
+                Preconditions.checkNotNull(reservedAllocationIdsRetriever);
     }
 
     @Override
     public SharedSlotProfileRetriever createFromBulk(Set<ExecutionVertexID> bulk) {
-        Set<AllocationID> allPriorAllocationIds =
-                bulk.stream()
-                        .map(priorAllocationIdRetriever)
-                        .filter(Objects::nonNull)
-                        .collect(Collectors.toSet());
-        return new MergingSharedSlotProfileRetriever(allPriorAllocationIds, bulk);
+        return new MergingSharedSlotProfileRetriever(reservedAllocationIdsRetriever.get(), bulk);
```

Comment:
With this change, I think we will no longer be able to schedule the following topology: Assume we have two pipelined regions A and B with each two tasks. Moreover, we are processing bounded streams. So in order to schedule A and B we need in total 4 slots. Now assume that 2 slots disappear. One used by A and one used by B. Moreover, assume that we won't get any new slots. In this case, Flink won't be able to schedule neither A nor B because of the reserved slots. However, w/o the reserved slots, it should be able to run both regions one after another.

This is probably more of an academic problem tbh. To be fair if we had only 2 slots in total, then we would have already failed if we hit the slot request timeout when scheduling the job initially. Ideally, though, we would eventually give up reserved slots if this allowed us to make progress.

Reply:
On a related note, maybe in the future we need to derive the `slotWillBeOccupiedIndefinitely` property from the boundedness of the input streams of an operator.

Reply:
Yes it is a limitation of local recovery on bounded streaming jobs. At the moment, I think the change is acceptable, because
- the problem has been there since local recovery was introduced. (although disappears when the pipelined region scheduling was introduced in 1.12, because local recovery restrictions can be broken in these problematic cases)
- slot request timeout will happen even if it is a bounded streaming job at the moment. So it is expected to have enough slots for all tasks to run at the same time, even for a bounded streaming jobs.

Later we can improve this for bounded streaming jobs,
1. first, we need to identify bounded streaming jobs. To achieve this, as you mentioned, we need to derive the `slotWillBeOccupiedIndefinitely` property from operators
2. if it is bounded streaming jobs, when allocating slots, it can try going the local recovery path first so that local recovery can still work for it in most cases. If slot allocation fails due to not enough slots, while a global check shows that there are enough slots for some of the regions to run, scheduler can clean up the reserved allocations before re-scheduling tasks.